File: Python/Mergesort.py
"""
Mergesort -> divide and conquer!
We divide our list in multiple smaller lists.
Each of those smaller lists will then be sorted.
At the end we merge the smaller, sorted lists to one big sorted list.
Worst case runtime O(n*log(n))
"""


def merge_sort(arr):
    # recursive implementation
    def merge(l, r):
        l_index, r_index = 0, 0
        result = []
        while l_index < len(l) and r_index < len(r):
            # compare the left and right elements
            # append the smaller one first to the new list
            if l[l_index] < r[r_index]:
                result.append(l[l_index])
                l_index += 1
            else:
                result.append(r[r_index])
                r_index += 1
        # either left or right has been used up, add the remaining elements of the other
        result += l[l_index:]
        result += right[r_index:]
        return result

    if len(arr) <= 1:
        return arr
    else:
        # split the array in 2 arrays recursively
        right = merge_sort(arr[len(arr)//2:])
        left = merge_sort(arr[:len(arr)//2])
        return merge(left, right)


if __name__ == "__main__":
    test_arr = [
        3,
        1,
        7,
        4,
        12,
        9,
        11,
        102
    ]

    print(merge_sort(test_arr))

File: C++/Mergesort.cpp
#include <iostream>
#include <vector>

#include "Randomdata.cpp"

void merge(std::vector<int> &arr, std::vector<int> &left, std::vector<int> &right){
    int left_idx = 0, right_idx = 0, arr_idx = 0;

    while (left_idx < left.size() && right_idx < right.size()) 
    {
        if (left[left_idx] <= right[right_idx]) {
            arr[arr_idx] = left[left_idx];
            left_idx++;
        }
        else {
            arr[arr_idx] = right[right_idx];
            right_idx++;
        }
        arr_idx++;
    }
    // Copy the remaining elements, if any
    while (left_idx < left.size()) {
        arr[arr_idx] = left[left_idx];
        left_idx++; arr_idx++;
    }
    while (right_idx < right.size()) {
        arr[arr_idx] = right[right_idx];
        right_idx++; arr_idx++;
    }
}

void mergeSort(std::vector<int> &arr){
    if (arr.size() <= 1){
        return;
    }
    int n = arr.size();
    int mid = n / 2;
    std::vector<int> left;
    std::vector<int> right;
    //split our unsorted array in 2 arrays
    for (int i = 0; i < mid; i++){
        left.push_back(arr[i]);
    }
    for (int i = 0; i < n - mid; i++){
        right.push_back(arr[mid + i]);
    }
    // recursively call mergeSort() untill size <= 1 
    mergeSort(left);
    mergeSort(right);
    merge(arr, left, right); // merge the subarrays back to our main array

}

void printArray(const std::vector<int>& arr){
    for (auto const& v: arr){
        std::cout << v << " | ";
    }
}

int main(){
    std::vector<int> arr = {12, 321, 12, 1, 5, 34, 27};
    mergeSort(arr);
    printArray(arr);

    std::cout << "\n\n";
    std::vector<int> bench = randomData(5000);
    printArray(bench);
    std::cout << "\n\n";
    mergeSort(bench);
    printArray(bench);

    return 0;
}